socketcan/libcanard/src/socketcan.c

```diff
@@ -162,43 +179,104 @@ int16_t socketcanPop(const SocketCANFD       fd,
+        // We use the CAN FD struct regardless of whether the CAN FD socket option is set.
+        // Per the user manual, this is acceptable because they are binary compatible.
+        struct iovec       iov           = {0};                    // Scatter/gather array items struct.
+        struct canfd_frame sockcan_frame = {0};                    // CAN FD frame storage.
+        iov.iov_base                     = &sockcan_frame;         // Starting address.
+        iov.iov_len                      = sizeof(sockcan_frame);  // Number of bytes to transfer.
+
+        // Determine the size of the ancillary data and zero-initialize the buffer for it.
+        // We require space for both the receive message and the time stamp.
+        // TODO: Aligned storage is used. --> Don't know how to do this in a portable way....
+        // TODO: C complains that the control_size is not known at compile time, hence the use of malloc/free.
+        //       Should not be needed, as those structs are fixed-size?? Ah, cmsghdr has a flexarray extension in c99?
+        const size_t control_size = sizeof(struct cmsghdr) + sizeof(struct timeval);
+        uint8_t*     control      = (uint8_t*) malloc(control_size);  // TODO: try to get rig of malloc!
+        if (control == NULL)
         {
-            return getNegatedErrno();
+            return -ENOMEM;
         }
+        (void) memset(control, 0, control_size);
 
-        // We use the CAN FD struct regardless of whether the CAN FD socket option is set.
-        // Per the user manual, this is acceptable because they are binary compatible.
-        struct canfd_frame cfd;
-        const ssize_t      read_size = read(fd, &cfd, sizeof(cfd));
-        if (read_size < 0)
+        // Initialize the message header used by recvmsg.
+        struct msghdr msg  = {0};           // Message header struct.
+        msg.msg_iov        = &iov;          // Scatter/gather array.
+        msg.msg_iovlen     = 1;             // Number of elements in the scatter/gather array.
+        msg.msg_control    = control;       // Ancillary data.
+        msg.msg_controllen = control_size;  // Ancillary data buffer length.
+
+        // Non-blocking receive messages from the socket and validate.
+        const ssize_t read_size = recvmsg(fd, &msg, MSG_DONTWAIT);
+        if (read_size <= 0)
         {
-            return getNegatedErrno();
+            const int neg_errno = getNegatedErrno();
+
+            // Return either the negated error code, or zero in case the socket is marked nonblocking and the receive
+            // operation would block. POSIX.1-2001 allows either EWOULDBLOCK or EAGAIN to be returned for this case,
+            // and does not require these constants to have the same value. Thus, we check for both possibilities.
+            free(control);
+            return (read_size < 0 && (neg_errno == -EWOULDBLOCK || neg_errno == -EAGAIN)) ? 0 : neg_errno;
```

Comment:
If the `poll(...)` says that there is data waiting to be read but `recvmsg(...)` says EWOULDBLOCK/EAGAIN, you have something sketchy going on and you should perhaps return an error. In the original code, it is handled by the next condition right below this one. I recommend you replace `if (read_size <= 0)` with `if (read_size < 0)` (strictly less) and remove the comment about the non-blocking case.

Reply:
Reviewing the logic, your proposal makes sense. Adjusted accordingly.